Fix Schedule Section Colors

dsnsgithub opened the pull request:
## Summary
Fixes #1321, Reverts https://github.com/icssc/AntAlmanac/commit/68f49dcf0c761a434d5997a5e3e67a91cb2d05a2

Restores/improves schedule section colors.

The root cause of the problem was https://github.com/icssc/AntAlmanac/pull/548:
```js
Math.round((color.l + delta) * 100 % 100)/100
```
when generating a new close color. This would keep pushing the luminance value higher until it would rollover, creating extremely bright and dark section colors.

---

The tweaking of HSL seems hacky, especially considering that luminance does not accurately represent how the human eye perceives light. 

I have changed the color generation so it uses MUI colors only. This will also make it significantly easier to add custom themes in the future.

<table>
  <tr>
    <th>Old</th>
<th>
    
<img height="400" alt="Schedule-old" src="https://github.com/user-attachments/assets/90b58fb7-c2fa-4288-b49e-64d1e105327b" />

</th>
  </tr>
  <tr>
     <th>New</th>
<th>
<img height="400" alt="Schedule-new" src="https://github.com/user-attachments/assets/9a7e3739-0dac-41db-9fb8-8f295b35c49b" />

</th>
  </tr>
</table>


In my testing, one other difference is that if the user picks a custom color, any other sections in the same course that are added will also share the same custom color, instead of a close color. I believe this is for the better as the luminance tweaking inherently creates UI/accessibility problems.


My code also has the added benefit of keeping the color consistent when re-adding the same classes.
## Test Plan
Refer to reproduction in #1321
## Issues

Closes #1321

## Future Followup
https://github.com/icssc/AntAlmanac/pull/1326 is likely not necessary.

alexespejo committed:
Styling: Schedule item colors (#1268)

Co-authored-by: vicksey <vickyzhang@hotmail.ca>

dsnsgithub committed:
fix: luminance rollover

- Check if a color provides enough contrast before generating it.

dsnsgithub committed:
Fix: Use correct variable name for darker color

alexespejo commented on apps/antalmanac/src/stores/scheduleHelpers.ts:
```diff
@@ -116,9 +150,22 @@ function colorIsContained(color: HSLColor, usedColors: Iterable<HSLColor>, delta
+ * @param maximum_luminance number: The maximum luminance value to consider the color visible, defaults to 0.8.
+ * @returns boolean: True if the color is visible enough, false otherwise.
+ */
+function isColorVisible(color: HSLColor, minimum_luminance = 0.2, maximum_luminance = 0.8): boolean {
```
formatting: camel case

dsnsgithub replied:
code was removed with later commits, no longer an issue

dsnsgithub committed:
Refactor: Improve color generation logic

Use predefined color variants for better contrast and avoid calculating
HSL colors and checking visibility.

dsnsgithub committed:
Bring color variants closer to original

alexespejo reviewed (changes requested):
Everything lgtm!

![image](https://github.com/user-attachments/assets/b6e693bf-aa4c-4a80-8ca6-dde90ee4a453)

alexespejo commented on apps/antalmanac/src/stores/scheduleHelpers.ts:
```diff
@@ -1,151 +1,80 @@
-}
-
-const defaultColors = [blue[500], pink[500], purple[500], green[500], amber[500], deepPurple[500], deepOrange[500]];
+const colorVariants: Record<string, string[]> = {
```
> I’ve never seen it — but I wouldn’t be surprised? I don’t think we should assume it will be at max 3 - Kevin

Honestly I assumed there was a max of 3 but he's not wrong. I think it's best to be dynamic with variants. In the future we might leverage the color picker to support something like async classes (not a great example)

dsnsgithub replied:
This code already handles extra section categories by falling back to the default section color, which seems to be the logical choice as it reuses colors when all color families are taken anyway. 

In my opinion, changing the colors of the sections dynamically is not ideal from a UI perspective. My perspective is that dynamically generating close colors by tweaking raw HSL/color values inherently creates UI inconsistencies, accessibility problems, and strays away from MUI. This dynamic solution doesn't fully fix the problem either, as the possible colors are still finite due to a minimum amount of variance between generated colors. By limiting the number of colors that can be used, we can vet the colors to ensure that text is visible while also using colors chosen by the MUI designers to be consistent with the rest of the UI.

It makes sense to me to add more colors as a precaution (even if the code can fallback), but they should fit the MUI theme and be vetted. Dynamic colors also means that it will not only stray away from MUI, but also violate the color choices of any potential themes that we decide to add in the future.


I can add the [400] and maybe even [500] series of colors just in case, and they would be also used when the user adds more than seven classes. Due to the way the code generates colors, the first looped color will be the default color, so it would reuse [300], then [400] and [500].

alexespejo replied:
I'm unopposed to defaulting back to the original colors if there is a scenario where we need more than 3. But I know it was intentional for the colors to desaturate for every subsection. 

cc: @KevinWu098

dsnsgithub replied:
We have three decent options for color order:
1. Revert to `[300], [200], [100]`
2. `[400], [300], [200], [100]` or `[500], [400], [300], [200], [100]`
3. Keep the same `[300], [200], [100], [400], [500]`

At [500] and at [400] for pink/purple, the text turns white for contrast, which IMO we should try to avoid.
The starting [400] series looks subjectively worse because we lose the pastel colors.
The [50] series of colors is objectively too bright.

I think we should just revert to `[300], [200], [100]` and fallback to [300] for any hypothetical new sections.

alexespejo replied:
Imma @vicksey just for a second opinion. Rotating the colors isn't a bad idea after all they are just hypothetical

alexespejo replied:
Now that I look at it the ~~oops~~ app that I took inspiration from with this issue are consistent with their font colors

dsnsgithub committed:
add more colors to color variants

alexespejo reviewed (changes requested):
burh i keep forgetting that i'm supposed to hit submit review
